### PR TITLE
Feature 79 create sale in organization

### PR DESCRIPTION
**Pull request by ignatIgnatov**
This ticket got pretty big. 
Includes creating a sale through an organization, returning a resource, returning a product, etc. 
Since I've advanced quite a bit in time, I'm probably missing something... 
I'm open for corrections.

### CODE REVIEW

File: src/main/java/jewellery/inventory/controller/OrganizationSaleController.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import jewellery.inventory.dto.response.OrganizationSaleResponseDto;
+import jewellery.inventory.dto.response.ProductReturnResponseDto;
+import jewellery.inventory.dto.response.ResourceReturnResponseDto;
+import jewellery.inventory.service.OrganizationSaleService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/organizations")
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/controller/OrganizationSaleController.java]: could include sales here and omit it in the methods
```suggestion
@RequestMapping("/organizations/sales")
```

File: src/main/java/jewellery/inventory/mapper/ProductInOrganizationMapper.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/mapper/ProductInOrganizationMapper.java:1]: obsolete change?
to roll it back:
`git checkout origin/test src/main/java/jewellery/inventory/mapper/ProductInOrganizationMapper.java`

File: src/main/java/jewellery/inventory/mapper/SaleMapper.java [hunk 2]
```diff
@@ -88,14 +82,44 @@ private List<ProductResponseDto> mapAllProductsToResponse(Sale sale) {
 
   public List<PurchasedResourceQuantityResponseDto> mapAllResourcesToResponse(Sale sale) {
     List<PurchasedResourceQuantityResponseDto> resources = new ArrayList<>();
-    if (sale.getResources() != null) {
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/mapper/SaleMapper.java]: Are we no longer concerned about null safety for `sale.getResources()`?

File: src/main/java/jewellery/inventory/service/OrganizationSaleService.java [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    organizationService.validateCurrentUserPermission(
+        organizationService.getOrganization(sale.getOrganizationSeller().getId()),
+        OrganizationPermission.RETURN_RESOURCE);
+
+    returnResourceFromSaleToOrganization(sale, resourceToReturn);
+    removeResourceFromSale(sale, resourceToReturn);
+
+    deleteSaleIfProductsAndResourcesAreEmpty(sale);
+    logger.info("Resource returned successfully. Resource ID: {}", resourceId);
+    return saleService.validateSaleAfterReturnResource(sale, resourceToReturn);
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/OrganizationSaleService.java:114]: this is also a bit unintuitive name.. but it is what it is for now

File: src/main/java/jewellery/inventory/service/SaleService.java [hunk 4]
```diff
@@ -252,17 +254,21 @@ private void throwExceptionIfResourceIsNotOwned(SaleRequestDto saleRequestDto) {
...
-        PurchasedResourceInUser purchasedResourceInUser =
-            getPurchasedResourceInUser(resourceRequest);
-        resources.add(purchasedResourceInUser);
+      List<PurchasedResourceInUser> resources = new ArrayList<>();
+      if (saleRequestDto.getResources() != null) {
+        for (PurchasedResourceQuantityRequestDto resourceRequest : saleRequestDto.getResources()) {
+          PurchasedResourceInUser purchasedResourceInUser =
+              getPurchasedResourceInUser(resourceRequest);
+          resources.add(purchasedResourceInUser);
+        }
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/SaleService.java]: This looks a bit weird. Why are we checking `if (saleRequestDto.getResources() != null)` twice?

File: src/test/java/jewellery/inventory/helper/SaleTestHelper.java [hunk 5]
```diff
@@ -147,6 +194,23 @@ private static void calculateProductsDiscounts(Sale sale, SaleResponseDto dto) {
...
+              BigDecimal salePrice =
+                  Optional.ofNullable(product.getSalePrice()).orElse(BigDecimal.ZERO);
+              BigDecimal discount =
+                  Optional.ofNullable(product.getDiscount()).orElse(BigDecimal.ZERO);
+              dto.setTotalDiscountedPrice(dto.getTotalDiscountedPrice().add(salePrice));
+              dto.setTotalDiscount(
+                  (dto.getTotalDiscount().add(discount))
+                      .divide(salePrice, MathContext.DECIMAL128)
+                      .multiply(getBigDecimal("100")));
+            });
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/helper/SaleTestHelper.java]: This looks pretty similar to the inner part of `calculateOrganizationResourcesDiscounts`. Can we plug in the salePrice and discount as parameters and extract the highlighted logic into a function?

File: src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java [hunk 6]
```diff
@@ -104,6 +152,135 @@ void setUp() {
...
+    SaleRequestDto saleRequestDto =
+        getSaleInOrganizationRequestDto(
+            organizationSeller, buyer, productResponse, resourcesInOrganizationResponseDto);
+
+    ResponseEntity<OrganizationSaleResponseDto> saleResponse =
+        createSaleInOrganization(saleRequestDto);
+
+    ResponseEntity<ResourceReturnResponseDto> resourceReturnResponse =
+        createReturnResourceResponse(
+            saleRequestDto, Objects.requireNonNull(saleResponse.getBody()));
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java]: Remove empty lines in preparation to have clear separation:
preparation

execution

assertion

File: src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java [hunk 7]
```diff
@@ -104,6 +152,135 @@ void setUp() {
...
+    assertEquals(
+        saleResponse.getBody().getProducts().get(0).getId(),
+        productReturnResponse.getBody().getReturnedProduct().getId());
+    assertEquals(
+        saleResponse.getBody().getResources().get(0).getResourceAndQuantity().getResource().getId(),
+        resourceReturnResponse.getBody().getReturnedResource().getId());
+  }
+
+  @Test
+  void returnProductFromSaleInOrganizationSuccessfully() throws JsonProcessingException {
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java:193]: Same comment about separating preparation/execution/assertion

File: src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java [hunk 8]
```diff
@@ -473,4 +662,88 @@ private ResponseEntity<ResourceReturnResponseDto> getResourceReturnResponse(
...
+            getBaseResourceInOrganizationAvailabilityUrl(),
+            request,
+            ResourcesInOrganizationResponseDto.class);
+
+    assertEquals(HttpStatus.CREATED, response.getStatusCode());
+    return response.getBody();
+  }
+
+  @NotNull
+  private static SaleRequestDto getSaleInOrganizationRequestDto(
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java]: Should this be in a helper class?

File: src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java [hunk 9]
```diff
@@ -473,4 +662,88 @@ private ResponseEntity<ResourceReturnResponseDto> getResourceReturnResponse(
...
+
+  private Organization createOrganizationInDatabase(
+      OrganizationRequestDto testOrganizationRequest) {
+    ResponseEntity<Organization> response =
+        this.testRestTemplate.postForEntity(
+            getBaseOrganizationsUrl(), testOrganizationRequest, Organization.class);
+
+    assertEquals(HttpStatus.CREATED, response.getStatusCode());
+    return response.getBody();
+  }
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java:688]: We have a similar method in OrganizationCrudIntegrationTest -> `createOrganizationsWithRequest`

**Review by VladoKat — CHANGES_REQUESTED** [review]
Overall logic looks good. I am concerned about how big and repetitive our tests are getting. We need to invest some time into cleaning them but I guess it won't be in the scope of this ticket.

File: src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java (hunk 9 above)
Comment (ignatIgnatov) [reply src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java:688]: Here I think not to change anything. Later I think to make some integration test helpers to reuse methods from.

File: src/main/java/jewellery/inventory/mapper/SaleMapper.java (hunk 2 above)
Comment (VladoKat) [reply src/main/java/jewellery/inventory/mapper/SaleMapper.java]: This is still pending

**Review by VladoKat — CHANGES_REQUESTED** [review]
one pending change from the previous review

File: src/main/java/jewellery/inventory/mapper/SaleMapper.java (hunk 2 above)
Comment (ignatIgnatov) [reply src/main/java/jewellery/inventory/mapper/SaleMapper.java]: Done!

**Review by VladoKat — APPROVED** [review]

**Pull request merged**